### PR TITLE
docs: animation of appearance and disappearance benefit

### PR DESCRIPTION
**Pull request by Anna-MariiaPetryk**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: pages/src/benefits/benefit.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  static get instance() {
+    return new AnimateService();
+  }
+
+  configFor = (el: HTMLElement) => this._configMap.get(el);
+  /**
+  * Intersection observable callback
+  */
+  private onIntersect(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
+    entries.forEach((entry: any) => {
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:31]:
```suggestion
    entries.forEach((entry: IntersectionObserverEntry) => {
```

File: pages/src/benefits/benefit.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+interface Iconfig {
+  group?: boolean;
+  repeat?: boolean;
+  delete?: boolean;
+}
+
+class AnimateService {
+
+  private _markedElements: HTMLElement[] = [];
+  private _io = new IntersectionObserver(this.onIntersect.bind(this), options);
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:18]: Please, use `bind` decorator

File: pages/src/benefits/benefit.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  private _configMap = new WeakMap<HTMLElement, Iconfig>();
+
+  @memoize()
+  static get instance() {
+    return new AnimateService();
+  }
+
+  configFor = (el: HTMLElement) => this._configMap.get(el);
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:26]: Please, define as method of class. And define return type.

File: pages/src/benefits/benefit.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        this._markedElements.push(entry.target);
+        this.postponedAnimate();
+      }
+      if (!entry.isIntersecting && this.configFor(entry.target)?.repeat) {
+        entry.target.setAttribute('esl-animate', 'true');
+      }
+      if (!this.configFor(entry.target)?.repeat && entry.isIntersecting) {
+        observer.unobserve(entry.target);
+        this.configFor(entry.target)?.delete = true;
+      }
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:42]:
```suggestion
      const target = entry.target;
      const config = this.configFor(target);
      if (entry.isIntersecting) {
        this._markedElements.push(target);
        this.postponedAnimate();
        if (!config?.repeat) {
          observer.unobserve(target);
          config?.delete = true;
        }
      } else {
        if (config?.repeat) {
          target.setAttribute('esl-animate', 'true');
        }
      }
```

File: pages/src/benefits/benefit.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import {ready} from '../../../src/modules/esl-utils/decorators/ready';
+import {memoize} from '../../../src/modules/esl-utils/decorators/memoize';
+import {bind} from '../../../src/modules/esl-utils/decorators/bind';
+import {ESLBaseElement} from '../../../src/modules/esl-base-element/core';
+
+
+const options = {
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:7]:
```suggestion
const options: IntersectionObserverInit = {
```

File: pages/src/benefits/benefit.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import {ready} from '../../../src/modules/esl-utils/decorators/ready';
+import {memoize} from '../../../src/modules/esl-utils/decorators/memoize';
+import {bind} from '../../../src/modules/esl-utils/decorators/bind';
+import {ESLBaseElement} from '../../../src/modules/esl-base-element/core';
+
+
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:6]: Extra row

File: pages/src/benefits/benefit.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  private _markedElements: Element[] = [];
+  private _io = new IntersectionObserver(this.onIntersect, options);
+  private _configMap = new WeakMap<Element, Iconfig>();
+
+  @memoize()
+  static get instance() {
+    return new AnimateService();
+  }
+
+  configFor(el: Element): Iconfig | undefined {
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:28]: It makes sense to start methods from verbs.
The name itself explains what the method does. 
I suggest `getConfigFor()`

File: pages/src/benefits/benefit.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    return this._configMap.get(el);
+  }
+
+  /**
+  * Intersection observable callback
+  */
+  @bind
+  private onIntersect(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
+    console.log(entries);
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:37]: forgotten debug

File: pages/src/benefits/benefit.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    this._configMap.set(el, Object.assign({repeat: false, group: false, delete: false}, config));
+    this._io.observe(el);
+  }
+
+  public unsubscribe(el: HTMLElement) {
+    this._io.unobserve(el);
+  }
+}
+
+
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:85]: extra row

File: pages/src/benefits/benefits.less [hunk 10]
```diff
@@ -80,3 +80,13 @@
     }
   }
 }
+
```
Comment (dshovchko) [inline pages/src/benefits/benefits.less:83]: extra row

File: pages/src/benefits/benefit.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      }
+      this.configFor(el)?.delete && this._configMap.delete(el);
+    });
+    this._markedElements = [];
+  }
+
+  private postponedAnimate = () => setTimeout(() => this.makeAppear(), 100);
+
+  public subscribe(el: HTMLElement, config?: Iconfig) {
+    this._configMap.set(el, Object.assign({repeat: false, group: false, delete: false}, config));
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:76]: Its better to define `{repeat: false, group: false, delete: false}` as constant.

File: pages/src/benefits/benefit.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  private makeAppear() {
+    let counter = 0;
+    this._markedElements.forEach(el => {
+      if (this.configFor(el)?.group) {
+        setTimeout(() => el.setAttribute('esl-animate', 'false'), 100 * counter);
+        counter++;
+      } else {
+        el.setAttribute('esl-animate', 'false');
+      }
+      this.configFor(el)?.delete && this._configMap.delete(el);
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:68]:
```suggestion
      const config = this.configFor(el);
      if (config?.group) {
        setTimeout(() => el.setAttribute('esl-animate', 'false'), 100 * counter);
        counter++;
      } else {
        el.setAttribute('esl-animate', 'false');
      }
      config?.delete && this._configMap.delete(el);
```

File: pages/src/benefits/benefit.ts [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    });
+    this._markedElements = [];
+  }
+
+  private postponedAnimate = () => setTimeout(() => this.makeAppear(), 100);
```
Comment (dshovchko) [inline pages/src/benefits/benefit.ts:73]:
```suggestion
  private postponedAnimate() {
    setTimeout(() => this.makeAppear(), 100);
  }
```

File: pages/src/benefits/benefit.ts (hunk 7 above)
Comment (ala-n) [reply pages/src/benefits/benefit.ts:28]: Not necessary to rename in case we will have a context. but in the general situation, I agree.

**Comment by Anna-MariiaPetryk** [comment]
@dshovchko, @ala-n, thanks for all advice ))))

**Comment by ala-n** [comment]
Closing, due to be covered by the #632

**Pull request closed without merging**